Remove f000 from atmos rocoto tasks for replay cases

EricSinsky-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
<!-- This description will become the commit message for the PR-->
<!-- Please use this format for your description:

  Describe your changes. Focus on the *what* and *why*. The *how* will be evident from the changes. In particular, be sure to note any interface changes, such as command line syntax, that will need to be communicated to users.

  At the end of your description, please be sure to add the issue this PR solves using the word "Resolves". If there are any issues that are related but not yet resolved (including in other repos), you may use "Refs".

  Resolves #1234
  Refs #4321
  Refs NOAA-EMC/repo#5678
-->

The main purpose of this PR is to remove the f000 from atmosphere-related rocoto tasks when `REPLAY_ICS` is set to `YES`. In cases where `REPLAY_ICS` is `YES` and `OFFSET_START_HOUR `is greater than `0`, it becomes necessary to have the first lead hour set to `OFFSET_START_HOUR ` for the atmosphere-related rocoto tasks. For example, when `OFFSET_START_HOUR ` is set to `3`, then the minimum lead time for the atmos_prod and atmos_ensstat rocoto tasks needs to be 3 and the minimum lead time for the ocean_prod rocoto task needs to be 6 (assuming `FHOUT_OCN` is 6). This PR makes this rocoto workflow set up possible by removing 0 from fhrs for atmosphere-related tasks in `gefs_tasks.py` when replaying. 

This PR also moves where f000 is being removed for the ocean_prod and ice_prod tasks. The if-block that performs this f000 removal has been moved from `tasks.py` to `gefs_tasks.py` and `gfs_tasks.py`.

This PR is a prerequisite for PR #2755. 

Refs #2771 
Resolves #2771 

# Type of change
<!-- Delete all except one -->
- Bug fix (fixes something broken)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
<!-- Please list any test you conducted, including the machine.

Example:
- Clone and build on WCOSS
- Cycled test on Orion
- Forecast-only on Hera
-->

This PR has been tested on WCOSS2 using a C384 GEFS case with `REPLAY_ICS` set to `YES` and `OFFSET_START_HOUR `set to `3`. 

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

EricSinsky-NOAA committed:
Update linters.yaml

The linters.yaml has been updated so that shellnorms can run whenever a push occurs to EricSinsky-NOAA/develop.

EricSinsky-NOAA committed:
Revert changes in linters.yaml

The changes to linters.yaml have been reverted.

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Use FHMIN_GFS for atmos component in tasks.py

FHMIN_GFS is used to determine fhmin for atmos component in the tasks.py script. FHMIN is used to determine fhmin for other components.

EricSinsky-NOAA committed:
Use FHMIN_GFS in FV3_predet

FHMIN_GFS is used instead of FHMIN in FV3_predet.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/issue2771

EricSinsky-NOAA commented:
A C48_S2SWA_gefs CI test case has successfully completed. Also successfully ran a special case where `FHMIN_GFS=3`,  `FHMIN=0` and `REPLAY_ICS=YES`. Marking as ready for review.

aerorahul commented:
@EricSinsky-NOAA I sent an email discussing the options.  Let's chat offline.

EricSinsky-NOAA committed:
Revert "Use FHMIN_GFS in FV3_predet"

This reverts commit 235ee26e103a8cff57421507cba2bbf1cf5e9def.

EricSinsky-NOAA committed:
Use offset_start_hour as atmos fhmin

When REPLAY_ICS is set to YES, use offset_start_hour as the atmos fhmin.

EricSinsky-NOAA committed:
Simplify variable definitions

EricSinsky-NOAA committed:
Remove 0 from fhrs in gefs/gfs tasks for ocn/ice

Have 0 removed from fhrs in the gefs_task.py and gfs_tasks.py python scripts for ocean and ice components.

EricSinsky-NOAA committed:
Remove f000 task for atmos_prod when replaying

The f000 task for atmos_prod is removed when replaying. The f000 task is removed in gefs_tasks.py instead of tasks.py

EricSinsky-NOAA committed:
Remove f000 task for atmos_ensstat when replaying

The f000 task for atmos_ensstat is removed when replaying. The f000 task is removed in gefs_tasks.py instead of tasks.py.

EricSinsky-NOAA committed:
Add Neil Barton's bugfix in exglobal_stage_ic

A bugfix has been added in exglobal_stage_ic to ensure ${MEMDIR:3} is not intrepreted in base 8.

aerorahul reviewed (dismissed):
lgtm

EricSinsky-NOAA commented:
Thank you, @aerorahul. I ran the C48_S2SWA_gefs CI test case, which has successfully completed.

WalterKolczynski-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -270,6 +270,17 @@ def _atmosoceaniceprod(self, component: str):
         fhrs = self._get_forecast_hours('gefs', self._configs[config], component)
+
+        # when replaying, atmos component does not have fhr 0, therefore remove 0 from fhrs
+        local_config = self._configs[config].deepcopy()
+        is_replay = local_config['REPLAY_ICS']
```
I don't think we need to make a deep copy here. We're not changing the configs, just reading them. We can use the original for that.
```suggestion
        is_replay = self._configs[config]['REPLAY_ICS']
```

EricSinsky-NOAA replied:
Sure. Deep copy has been removed here and for atmos_ensstat.

WalterKolczynski-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -314,6 +325,13 @@ def atmos_ensstat(self):
         fhrs = self._get_forecast_hours('gefs', self._configs['atmos_ensstat'])
+
+        # when replaying, atmos component does not have fhr 0, therefore remove 0 from fhrs
+        local_config = self._configs['atmos_ensstat'].deepcopy()
+        is_replay = local_config['REPLAY_ICS']
```
Same here
```suggestion
        is_replay = self._configs['atmos_ensstat']['REPLAY_ICS']
```

EricSinsky-NOAA committed:
Remove deep copy according to reviewer's comment

Deep copy has been removed in gefs_tasks.py since no configs are being changed.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/issue2771

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/issue2771

aerorahul committed:
Merge branch 'develop' into feature/issue2771
